enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,356 @@
+|--------------------------------------------------	|-----------------------------------------  	|-------	                            |
+| The whole stack of components is needed         	| `microshift`                   	              | Nothing would change in this case, installing microshift, which becomes a virtual package, installs the necessary dependencies to help maintain backwards compatibility	|
+| Only advanced networking is needed                | `microshift-core`, `microshift-networking-ovnk`  |                                    	|
+| The bare minimum set of components is necessary   | `microshift-core`                             |                                   	|
+| Only persistent volumes are necessary             | `microshift-core`, `microshift-storage-topolvm` |                                     |
+
+
+The packages for networking and storage would be different in upstrean, containing
+references to unrestricted (OKD) container images, removing the need for a pull secret.
+
+The following changes would be introduced to allow the described type of modularity:
+
+Refactor and reuse the kustomize loader into a new module for core components which
+would load components like:
+
+`/usr/lib/microshift/components.d/004-multus/kustomize.yaml`
+`/usr/lib/microshift/components.d/005-networking-ovnk/kustomize.yaml`
+`/usr/lib/microshift/components.d/010-storage-topolvm/kustomize.yaml`
+
+  MicroShift component manager would scan the `/usr/lib/microshift/components.d/`
+  directory executing the component manifests in alphabetical order.
+
+  Today for the CNI/CSI components there is no wait or readiness conditions,
+  if this needed to be implemented in the future it could be defined in yaml
+  metadata.
+
+  `/etc/microshift/components.d/` would also be provided to allow passing down
+  specific configuration down to components; for example:
+
+  `/etc/microshift/components.d/010-storage-topolvm/config.yaml`
+
+   such yaml would parsed and passed down on the templates of the following
+   point as `{{ .config.value }}` or `{{ .config.sub.value }}`.
+
+*  Kustomize yamls would be pre-processed with go templates to allow insertion of
+  variables as we do today for core assets, i.e.: `{{ .KubeconfigDir }}`,
+  `{{ .BaseDomain }}`,
+  [etc...](https://github.com/openshift/microshift/blob/eb0488fd167333256c69f4eeedc5463c4035a532/pkg/components/render.go#L24).
+  Image references will be injected in the `images` section
+  of the kustomize format not being embedded in `microshift/pkg/releases.go`
+  or passed down as template variables anymore.
+
+* The functionality used in the topo-lvm component to embed the lvmd config file
+  in a config map could be performed with a config map generator, which would contain
+  a default config file. The default config file could be overriden via a custom component
+  variable.
```

Comment:
What is a "custom component variable"? Why not assume users replace the config file that gets embedded into the ConfigMap by the generator?